apifile.go

```diff
@@ -0,0 +1,254 @@
+package httpapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/interface-go-ipfs-core"
+)
+
+const forwardSeekLimit = 1 << 14 //16k
+
+func (api *UnixfsAPI) Get(ctx context.Context, p iface.Path) (files.Node, error) {
```

Comment:
> 2. Start with get but switch to this iterative approach if we skip a file

I was thinking about this quite a bit before settling for the current solution. There are some problems:
* If user skips entry, we need to have a list of hashes he has already seen, because we want to call `ls --stream`
* (might be wrong) `ipfs get` can take a while to start on large directories (it uses `unixfsDir.ForEachLink` under the hood, I'm not sure how that works for hamt)

If we're ok with the skip logic being rather expensive, I can try implementing it.

The "browsing" option makes a lot of sense, but it's a bit hard to fit into go-ipfs implementation, which just works for everything. We could add checks there to make sure that the behaviour is at least consistent across implementations.

> 1. Won't play well with sessions

I was thinking about a global flag which would take random strings (UUIDs or similar), go to some map, check if session with that ID exists, if not create it, and use that (https://github.com/ipfs/go-ipfs/issues/5994).

Reply:
> (might be wrong) ipfs get can take a while to start on large directories (it uses unixfsDir.ForEachLink under the hood, I'm not sure how that works for hamt)

In-order.

> If we're ok with the skip logic being rather expensive, I can try implementing it.

Let's just benchmark against some _local_ dataset and see if it makes a difference. Regardless, we don't have to block merging this on optimizing it.

> I was thinking about a global flag which would take random strings (UUIDs or similar), go to some map, check if session with that ID exists, if not create it, and use that (ipfs/go-ipfs#5994).

:+1: although I'm not sure how to garbage collect them.

Reply:
https://github.com/ipfs/go-ipfs-http-client/issues/3